ci/Dockerfile

```diff
@@ -14,39 +14,97 @@ ARG parallelism=18
-spack compiler find
+  # Add package repositories
+  spack --timestamp repo add https://github.com/FNALssi/fnal_art.git
+  spack --timestamp repo add https://github.com/Framework-R-D/phlex-spack-recipes.git
 
-# Discover externals
-# - Python is excluded because the python provided by the system does
-#   not necessarily include the Python.h header file.  We therefore
-#   need Spack to build Python.
-spack external find --exclude python
+  # Discover compilers
+  spack --timestamp compiler find
+
+  # Discover externals
+  # - Python is excluded because the python provided by the system does
+  #   not necessarily include the Python.h header file.  We therefore
+  #   need Spack to build Python.
+  spack --timestamp external find --exclude python
+EOF
+
+RUN <<EOF
+  # Configure a buildcache if mounted at build-time
+  if [ -d "/build-cache" ]; then
+    . spack/share/spack/setup-env.sh
+    spack --timestamp mirror add --scope=site --type binary \
+      scisoft_phlex_ci /build-cache
+  fi
 EOF
 
 COPY ./spack.yaml /
 
 RUN <<EOF
-# Setup environment
-. spack/share/spack/setup-env.sh
-
-# Create and prepare phlex-development environment
-spack env create phlex-ci spack.yaml
-spack env activate phlex-ci
+  # Create and prepare phlex-development environment
 
-# Display concretized solution
-spack concretize
+  . spack/share/spack/setup-env.sh
+  spack versions -s llvm | grep -qEe '^[[:space:]]+21\.1\.4$' || spack checksum -ab llvm 21.1.4
+  spack --timestamp env create phlex-ci spack.yaml
+  spack --timestamp env activate phlex-ci
 
-# Install almost everything
-spack install --fail-fast -j $parallelism -p 1 $(spack find -r --no-groups | sed -Ene 's&^ - &&p' | grep -v phlex)
+  # Display concretized solution
+  spack --timestamp concretize
 EOF
 
 COPY ./entrypoint.sh /
+
 RUN chmod +x /entrypoint.sh
+
+RUN <<EOF
+  # Install most of the tools as their own layer
+  . /entrypoint.sh
+  spack --timestamp install --fail-fast -j $parallelism -p 1 \
+    --no-add --only-concrete --no-check-signature \
+    cmake lcov ninja py-cmake-format py-gcovr
+  # Clean up to reduce layer size
+  spack clean -dfs
+EOF
+
+RUN <<EOF
+  # Install llvm as its own layer
+  . /entrypoint.sh
+  spack --timestamp install --fail-fast -j $parallelism -p 1 \
+    --no-add --only-concrete --no-check-signature llvm
+  # Clean up to reduce layer size
+  spack clean -dfs
+EOF
+
+RUN <<EOF
+  # Install almost everything else
+  . /entrypoint.sh
+  spack --timestamp install --fail-fast -j $parallelism -p 1 \
+    --no-add --only-concrete --no-check-signature \
+    $(spack find -r --no-groups | sed -Ene 's&^ - &&p' | grep -v phlex)
+  # Clean up to reduce layer size
+  spack clean -dfs
+EOF
+
+RUN <<EOF
+  # Install Phlex' dependencies
+  . /entrypoint.sh
+  spack --timestamp install --fail-fast -j $parallelism -p 1 \
+    --no-add --only-concrete --only dependencies --no-check-signature \
+    phlex
+  # Clean up to reduce layer size
+  spack clean -dfs
+EOF
```

Comment:
Why separate these two steps?  The first one appears to install everything that is not already installed; the second one will ensure that everything except `phlex` is installed.  The consequence is the same, no?